part2/stages/Set-password

```diff
@@ -18,93 +18,248 @@
+    sed -n "${first}p"  "${file}" | tr -d '\n' > "${pw1}"
+    local pw2=$( mktemp --tmpdir="${KEYDIR}" )
+    sed -n "${second}p" "${file}" | tr -d '\n' > "${pw2}"
+
+    if ! diff -q "${pw1}" "${pw2}" 1>/dev/null 2>&1 ; then
+        rm "${pw1}" "${pw2}"
+        return ${ERR_MISMATCH}
     fi
 
-    while [ 1 ] ; do
-        show_cursor
+    if [ ! -s "${pw1}" ] ; then
+        rm "${pw1}" "${pw2}"
+        return ${ERR_EMPTY}
+    fi
 
-        CAPTURE=$(mktemp -t)
-        dialog --cancel-label Previous --insecure --passwordbox "Please enter a new password for the system" 0 50 "" 2>${CAPTURE}
+    cp "${pw1}" "${out}"
+    rm "${pw1}" "${pw2}"
 
-        OPT="$?"
-        if [ ${OPT} != 0 ] ; then
-            rm -f ${CAPTURE}
-            [ ${OPT} != 1 ]   || exit ${Previous}
-            [ ${OPT} != 255 ] || exit ${Previous}
-            exit ${Abort}
+    return 0
+}
+
+# Expects a file of 4 or 6 lines:
+# admin password
+# admin password re-type
+# recovery password
+# recovery password re-type
+#  and if MEASURE_LAUNCH=true
+# tpm password
+# tpm password re-type
+parse_passwords() {
+    local file="$1"
+    local ret=0
+    local ret2=0
+    local rows=4
+    [ "${MEASURE_LAUNCH}" = "true" ] && rows=6
+    if [ "$( wc -l ${file} | awk '{ print $1 }' )" -ne ${rows} ]; then
+        return 1;
+    fi
+
+    if [ ! -e "${PASSWORD_HASH}" ]; then
+        check_and_save_password "${file}" 1 "${admin_pass}"
+        ret=$?
+        if [ $ret -ne 0 ] ; then
+            error_dialog "Admin" $ret
+            ret2=1
+        else
+            openssl passwd -1 -in "${admin_pass}" | tr -d '\n' > "${PASSWORD_HASH}"
+            rm -f "${admin_pass}"
         fi
+    fi
 
-        REPEAT=$(mktemp -t)
-        dialog --cancel-label Previous --insecure --passwordbox "Please enter the same password to confirm" 0 50 "" 2>${REPEAT}
+    if [ ! -e "${recov_pass}" ]; then
+        check_and_save_password "${file}" 3 "${recov_pass}"
+        ret=$?
+        if [ $ret -ne 0 ] ; then
+            error_dialog "Recovery" $ret
+            ret2=1
+        fi
+    fi
 
-        OPT="$?"
-        if [ ${OPT} != 0 ] ; then
-            rm -f ${CAPTURE}
-            rm -f ${REPEAT}
-            [ ${OPT} != 1 ]   || exit ${Previous}
-            [ ${OPT} != 255 ] || exit ${Previous}
-            exit ${Abort}
+    if [ ! -e "${tpm_pass}" ] ; then
+        check_and_save_password "${file}" 5 "${tpm_pass}"
+        ret=$?
+        if [ $ret -ne 0 ] ; then
+            error_dialog "TPM" $ret
+            ret2=1
         fi
+    fi
 
-        PLAIN_TEXT_PASSWORD=$(cat ${CAPTURE})
+    echo ${ret2}
+    return ${ret2}
+}
 
-        if [ -z "${PLAIN_TEXT_PASSWORD}" ] ; then
-            rm -f ${CAPTURE}
-            rm -f ${REPEAT}
-            hide_cursor
-            dialog --colors --ok-label Continue --msgbox '  \ZbERROR:\ZB Empty password!\n\n    Please try again.' 7 30
-            continue
+query_passwords() {
+    while [ 1 ] ; do
```

Comment:
`[ 1 ]` will always be true, just like `[ 0 ]` though...
Maybe just use `true`?